File: dsa-problems/LinkedLists/solutions/python/LNK-014-robotics-palindrome-one-skip.py
import sys

class ListNode:
    def __init__(self, val=0):
        self.val = val
        self.next = None

def is_palindrome(vals, left, right):
    while left < right:
        if vals[left] != vals[right]:
            return False
        left += 1
        right -= 1
    return True

def can_be_palindrome(head: ListNode) -> bool:
    vals = []
    curr = head
    while curr:
        vals.append(curr.val)
        curr = curr.next
        
    left, right = 0, len(vals) - 1
    while left < right:
        if vals[left] != vals[right]:
            return is_palindrome(vals, left + 1, right) or \
                   is_palindrome(vals, left, right - 1)
        left += 1
        right -= 1
        
    return True

def main():
    input = sys.stdin.read
    data = input().split()
    if not data:
        return
    
    iterator = iter(data)
    try:
        n = int(next(iterator))
        dummy = ListNode()
        cur = dummy
        for _ in range(n):
            cur.next = ListNode(int(next(iterator)))
            cur = cur.next
            
        print("true" if can_be_palindrome(dummy.next) else "false")
    except StopIteration:
        pass

if __name__ == "__main__":
    main()

File: dsa-problems/LinkedLists/solutions/javascript/LNK-014-robotics-palindrome-one-skip.js
const readline = require("readline");

class ListNode {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

function isPalindrome(vals, left, right) {
  while (left < right) {
    if (vals[left] !== vals[right]) {
      return false;
    }
    left++;
    right--;
  }
  return true;
}

function canBePalindrome(head) {
  const vals = [];
  let curr = head;
  while (curr) {
    vals.push(curr.val);
    curr = curr.next;
  }

  let left = 0;
  let right = vals.length - 1;

  while (left < right) {
    if (vals[left] !== vals[right]) {
      return isPalindrome(vals, left + 1, right) || 
             isPalindrome(vals, left, right - 1);
    }
    left++;
    right--;
  }
  return true;
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let data = [];
rl.on("line", (line) => data.push(...line.trim().split(/\s+/).filter(x => x)));
rl.on("close", () => {
  if (data.length === 0) return;
  let idx = 0;
  const n = parseInt(data[idx++], 10);
  
  const dummy = new ListNode(0);
  let cur = dummy;
  for (let i = 0; i < n; i++) {
    cur.next = new ListNode(parseInt(data[idx++], 10));
    cur = cur.next;
  }

  console.log(canBePalindrome(dummy.next) ? "true" : "false");
});
